File: webapp/core/permissions.py
from __future__ import annotations

from typing import Any

from core.modules import MODULES, Module, get_function, get_module


PermMap = dict[tuple[str, str], tuple[bool, bool]]


def is_admin(user: dict[str, Any] | None) -> bool:
    return bool(user and user.get("role_id") == "admin")


def _flag(user: dict[str, Any], module_id: str, function_id: str, write: bool) -> bool:
    if is_admin(user):
        return True
    perms: PermMap = user.get("perms") or {}
    read, can_w = perms.get((module_id, function_id), (False, False))
    return bool(can_w if write else read or can_w)


def can_read(user: dict[str, Any], module_id: str, function_id: str | None = None) -> bool:
    if is_admin(user):
        return True
    if function_id:
        return _flag(user, module_id, function_id, write=False)
    module = get_module(module_id)
    if not module:
        return False
    return any(_flag(user, module_id, fn.id, write=False) for fn in module.functions)


def can_write(user: dict[str, Any], module_id: str, function_id: str | None = None) -> bool:
    if is_admin(user):
        return True
    if function_id:
        return _flag(user, module_id, function_id, write=True)
    module = get_module(module_id)
    if not module:
        return False
    return any(_flag(user, module_id, fn.id, write=True) for fn in module.functions)


def readable_functions(user: dict[str, Any], module: Module):
    return tuple(fn for fn in module.functions if can_read(user, module.id, fn.id))


def nav_modules(user: dict[str, Any]) -> list[dict[str, Any]]:
    items = []
    for module in MODULES:
        fns = readable_functions(user, module)
        if not fns:
            continue
        items.append({"module": module, "functions": fns})
    return items


def permission_grid(stored: PermMap) -> list[dict[str, Any]]:
    rows = []
    for module in MODULES:
        functions = []
        for fn in module.functions:
            read, write = stored.get((module.id, fn.id), (False, False))
            functions.append(
                {
                    "id": fn.id,
                    "title": fn.title,
                    "key": f"{module.id}:{fn.id}",
                    "can_read": bool(read or write),
                    "can_write": bool(write),
                }
            )
        rows.append({"module": module, "functions": functions})
    return rows


def parse_permission_form(read_keys: list[str], write_keys: list[str]) -> PermMap:
    writes = set(write_keys)
    reads = set(read_keys) | writes
    out: PermMap = {}
    for key in reads | writes:
        if ":" not in key:
            continue
        module_id, function_id = key.split(":", 1)
        if not get_function(module_id, function_id):
            continue
        write = key in writes
        out[(module_id, function_id)] = (True, write)
    return out

File: backend/src/permissions.js
const MODULES = require('./modules.json');

const ACTION_IDS = [
    "read", "create", "write", "delete", "submit", "cancel",
    "amend", "print", "email", "export", "import", "report", "share"
];

const STANDARD_WRITE = new Set([
    "read", "create", "write", "submit", "cancel", "amend",
    "print", "email", "export", "report", "share"
]);

function emptyFlags() {
    const flags = {};
    ACTION_IDS.forEach(action => flags[action] = false);
    return flags;
}

function accessFlags({ write = false, full = false } = {}) {
    if (full) {
        const flags = {};
        ACTION_IDS.forEach(action => flags[action] = true);
        return flags;
    }
    const flags = emptyFlags();
    flags.read = true;
    if (write) {
        for (let action of STANDARD_WRITE) {
            flags[action] = true;
        }
    }
    return flags;
}

function flagsFromLegacy(read, write) {
    if (write) return accessFlags({ write: true });
    if (read) return accessFlags({ write: false });
    return emptyFlags();
}

function normalizeFlags(value) {
    const flags = emptyFlags();
    if (value && typeof value === 'object' && !Array.isArray(value)) {
        ACTION_IDS.forEach(action => {
            flags[action] = !!value[action];
        });
        return flags;
    }
    if (Array.isArray(value) && value.length >= 2) {
        return flagsFromLegacy(!!value[0], !!value[1]);
    }
    return flags;
}

function isAdmin(user) {
    return !!(user && user.role_id === 'admin');
}

function _flagsFor(user, moduleId, functionId) {
    const perms = user?.perms || {};
    // Note: Python was using a tuple key `(module_id, function_id)`. We need to adapt it.
    // Assuming user.perms is passed from the database correctly.
    // In db.py, perms is returned as dict with tuple keys. In JSON we'll use "moduleId:functionId".
    const key = `${moduleId}:${functionId}`;
    return normalizeFlags(perms[key]);
}

function getModule(moduleId) {
    return MODULES.find(m => m.id === moduleId);
}

function canDo(user, moduleId, functionId, action) {
    if (isAdmin(user)) return true;
    if (functionId) {
        const flags = _flagsFor(user, moduleId, functionId);
        return !!flags[action];
    }
    const module = getModule(moduleId);
    if (!module) return false;
    return module.functions.some(fn => canDo(user, moduleId, fn.id, action));
}

function canRead(user, moduleId, functionId = null) {
    return canDo(user, moduleId, functionId, "read");
}

function canWrite(user, moduleId, functionId = null) {
    return canDo(user, moduleId, functionId, "write");
}

function navModules(user) {
    const nav = [];
    for (const m of MODULES) {
        if (canRead(user, m.id)) {
            // Clone the module so we don't mutate the global MODULES array
            const clonedModule = { ...m };
            if (clonedModule.functions) {
                clonedModule.functions = clonedModule.functions.filter(fn => canRead(user, m.id, fn.id));
            }
            nav.push(clonedModule);
        }
    }
    return nav;
}

module.exports = {
    MODULES,
    isAdmin,
    canDo,
    canRead,
    canWrite,
    navModules,
};
